### ISSUE TITLE
bundle install does not complete

### ISSUE THREAD

**Issue by sonalkr132** [rubygems/rubygems#4240]
### Describe the problem as clearly as you can

bundle install gets stuck on `Resolving dependencies.....` step.

### Post steps to reproduce the problem

* ruby: 2.5.8p224
* gem: 3.2.4
* bundler: 2.2.4

<details><summary>Gemfile</summary>

```ruby
source ENV['GEM_SOURCE'] || 'https://rubygems.org'

def location_for(place_or_version, fake_version = nil)
  git_url_regex = %r{\A(?<url>(https?|git)[:@][^#]*)(#(?<branch>.*))?}
  file_url_regex = %r{\Afile:\/\/(?<path>.*)}

  if place_or_version && (git_url = place_or_version.match(git_url_regex))
    [fake_version, { git: git_url[:url], branch: git_url[:branch], require: false }].compact
  elsif place_or_version && (file_url = place_or_version.match(file_url_regex))
    ['>= 0', { path: File.expand_path(file_url[:path]), require: false }]
  else
    [place_or_version, { require: false }]
  end
end

ruby_version_segments = Gem::Version.new(RUBY_VERSION.dup).segments
minor_version = ruby_version_segments[0..1].join('.')

group :development do
  gem "fast_gettext", '1.1.0',                                   require: false if Gem::Version.new(RUBY_VERSION.dup) < Gem::Version.new('2.1.0')
  gem "fast_gettext",                                            require: false if Gem::Version.new(RUBY_VERSION.dup) >= Gem::Version.new('2.1.0')
  gem "json_pure", '<= 2.0.1',                                   require: false if Gem::Version.new(RUBY_VERSION.dup) < Gem::Version.new('2.0.0')
  gem "json", '= 1.8.1',                                         require: false if Gem::Version.new(RUBY_VERSION.dup) == Gem::Version.new('2.1.9')
  gem "json", '= 2.0.4',                                         require: false if Gem::Requirement.create('~> 2.4.2').satisfied_by?(Gem::Version.new(RUBY_VERSION.dup))
  gem "json", '= 2.1.0',                                         require: false if Gem::Requirement.create(['>= 2.5.0', '< 2.7.0']).satisfied_by?(Gem::Version.new(RUBY_VERSION.dup))
  gem "rb-readline", '= 0.5.5',                                  require: false, platforms: [:mswin, :mingw, :x64_mingw]
  gem "puppet-module-posix-default-r#{minor_version}", '~> 0.3', require: false, platforms: [:ruby]
  gem "puppet-module-posix-dev-r#{minor_version}", '~> 0.3',     require: false, platforms: [:ruby]
  gem "puppet-module-win-default-r#{minor_version}", '~> 0.3',   require: false, platforms: [:mswin, :mingw, :x64_mingw]
  gem "puppet-module-win-dev-r#{minor_version}", '~> 0.3',       require: false, platforms: [:mswin, :mingw, :x64_mingw]
  gem "beaker-rspec",                                            require: false
  gem "beaker-puppet",                                           require: false
  gem "beaker-docker",                                           require: false
  gem "beaker-puppet_install_helper",                            require: false
  gem "beaker-module_install_helper",                            require: false
end

puppet_version = ENV['PUPPET_GEM_VERSION']
facter_version = ENV['FACTER_GEM_VERSION']
hiera_version = ENV['HIERA_GEM_VERSION']

gems = {}

gems['puppet'] = location_for(puppet_version)

# If facter or hiera versions have been specified via the environment
# variables

gems['facter'] = location_for(facter_version) if facter_version
gems['hiera'] = location_for(hiera_version) if hiera_version

if Gem.win_platform? && puppet_version =~ %r{^(file:///|git://)}
  # If we're using a Puppet gem on Windows which handles its own win32-xxx gem
  # dependencies (>= 3.5.0), set the maximum versions (see PUP-6445).
  gems['win32-dir'] =      ['<= 0.4.9', require: false]
  gems['win32-eventlog'] = ['<= 0.6.5', require: false]
  gems['win32-process'] =  ['<= 0.7.5', require: false]
  gems['win32-security'] = ['<= 0.2.5', require: false]
  gems['win32-service'] =  ['0.8.8', require: false]
end

gems.each do |gem_name, gem_params|
  gem gem_name, *gem_params
end

# Evaluate Gemfile.local and ~/.gemfile if they exist
extra_gemfiles = [
  "#{__FILE__}.local",
  File.join(Dir.home, '.gemfile'),
]

extra_gemfiles.each do |gemfile|
  if File.file?(gemfile) && File.readable?(gemfile)
    eval(File.read(gemfile), binding)
  end
end
# vim: syntax=ruby
```
</details>

### Which command did you run?

`DEBUG_RESOLVER=1 bundle install --verbose --without system_tests`

### What were you expecting to happen?

bundle install to complete.

### What actually happened?

resolver seems to be stuck in a loop.
```
Requiring nested dependencies (addressable (~> 2.5), aws-sdk-ec2 (~> 1), CFPropertyList (~> 2.2), concurrent-ruby (~> 1.0), hiera-eyaml (~> 3), logging (~> 2.2), minitar (~> 0.6), net-scp (~> 1.2), net-ssh (>= 4.0), orchestrator_client (~> 0.4), puppet (< 7, >= 6.4.0), puppet-resource_api (>= 1.8.1), r10k (~> 3.1), ruby_smb (~> 1.0), terminal-table (~> 1.8), winrm (~> 2.0), winrm-fs (~> 1.3), Ruby (~> 2.3))
BUNDLER: Creating possibility state for Ruby (~> 2.3) (1 remaining)
BUNDLER:   Attempting to activate [Ruby (2.5.8.224) (ruby), Ruby (2.5.8.224) (x86_64-linux)]
BUNDLER:   Found existing spec ([Ruby (2.5.8.224) (ruby), Ruby (2.5.8.224) (x86_64-linux)])
BUNDLER:   Creating possibility state for puppet (< 7, >= 6.4.0) (52 remaining)
BUNDLER:     Attempting to activate [puppet (6.19.1) (x86_64-linux)]
BUNDLER:     Found existing spec ([puppet (7.1.0) (x86_64-linux)])
BUNDLER:     Unsatisfied by existing spec ([puppet (7.1.0) (x86_64-linux)])
BUNDLER:     Unwinding for conflict: puppet (< 7, >= 6.4.0) to 184
BUNDLER:   Creating possibility state for puppet (< 7, >= 6.4.0) (51 remaining)
BUNDLER:    Attempting to activate [puppet (6.19.1) (ruby)]
BUNDLER:    Found existing spec ([puppet (7.1.0) (x86_64-linux)])
BUNDLER:    Unsatisfied by existing spec ([puppet (7.1.0) (x86_64-linux)])
BUNDLER:    Unwinding for conflict: puppet (< 7, >= 6.4.0) to 184
BUNDLER:  Creating possibility state for puppet (< 7, >= 6.4.0) (50 remaining)
....
```
Originally reported at: https://help.rubygems.org/discussions/problems/38518-resolving-dependencies-never-finish
The issue doesn't exist on bundler 2.1,4.

**Comment by deivid-rodriguez** [comment]
Thanks for redirecting this to us @sonalkr132, it should get fixed by #4243.

**Comment by Dan33l** [comment]
thank you so much @sonalkr132 you made my day happier !
I was able to downgrade `bundler` to 2.1.4 and now it works.

And also thank you to author of this blog post :
https://dev.to/st0012/completely-remove-the-default-bundler-from-ci-environment-j0c
